### PR TITLE
Full CfL search by prediction error

### PR DESCRIPTION
**Pull request by barrbrain**
[AWCY results](https://beta.arewecompressedyet.com/?job=master-352a95563ec3ec1f6401590b7e384ad47524dbc6&job=rav1e-cfl-pred-full-2%402018-08-29T09%3A51%3A13.330Z) at default speed level:

|   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |    ---: |    ---: |     ---: |   ---: |    ---: |       ---: |
| 0.0110 | -7.0461 | -4.9412 |   0.0287 | 0.0321 |  0.0721 |    -2.2357 |

### CODE REVIEW

File: src/rdo.rs [hunk 1]
```diff
@@ -288,14 +288,6 @@ pub fn rdo_mode_decision(
...
         if !best_mode_chroma.is_intra() { continue; }
         let mut wr: &mut dyn Writer = &mut WriterCounter::new();
-        // these rules follow TX_MODE_LARGEST
-        let tx_size = match bsize {
-            BlockSize::BLOCK_4X4 => TxSize::TX_4X4,
-            BlockSize::BLOCK_8X8 => TxSize::TX_8X8,
-            BlockSize::BLOCK_16X16 => TxSize::TX_16X16,
-            _ => TxSize::TX_32X32
-        };
-        let tx_type = TxType::DCT_DCT;
```
Comment (barrbrain) [inline src/rdo.rs]: [AWCY results](https://beta.arewecompressedyet.com/?job=rav1e-cfl-pred-full%402018-08-29T09%3A19%3A18.651Z&job=rav1e-cfl-pred-full-2%402018-08-29T09%3A51%3A13.330Z) for fixing this mistake:

|   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |    ---: |    ---: |     ---: |   ---: |    ---: |       ---: |
| 0.0698 | -0.6611 | -0.2755 |   0.1140 | 0.0975 |  0.1082 |    -0.1782 |

File: src/rdo.rs [hunk 2]
```diff
@@ -283,51 +284,59 @@ pub fn rdo_mode_decision(
...
-      if chroma_mode != PredictionMode::UV_CFL_PRED {
-        cfl_alpha_set.truncate(1);
-      } else if !best_mode_chroma.is_intra() {
-        continue;
+      let mut cfl = CFLParams::new();
+      if chroma_mode == PredictionMode::UV_CFL_PRED {
+        if !best_mode_chroma.is_intra() { continue; }
+        let mut wr: &mut dyn Writer = &mut WriterCounter::new();
+        // these rules follow TX_MODE_LARGEST
+        let tx_size = match bsize {
```
Comment (ycho) [inline src/rdo.rs]: Oh no!  The (tx_size, tx_type) has been decided outside of chroma rdo loop!

File: src/rdo.rs [hunk 3]
```diff
@@ -283,51 +284,59 @@ pub fn rdo_mode_decision(
...
+            BlockSize::BLOCK_4X4 => TxSize::TX_4X4,
+            BlockSize::BLOCK_8X8 => TxSize::TX_8X8,
+            BlockSize::BLOCK_16X16 => TxSize::TX_16X16,
+            _ => TxSize::TX_32X32
+        };
+        let tx_type = TxType::DCT_DCT;
```
Comment (ycho) [inline src/rdo.rs]: Enforcing tx_type even harm since the best tx_type is already chosen above!

File: src/rdo.rs [hunk 4]
```diff
@@ -288,14 +288,6 @@ pub fn rdo_mode_decision(
       if chroma_mode == PredictionMode::UV_CFL_PRED {
         if !best_mode_chroma.is_intra() { continue; }
         let mut wr: &mut dyn Writer = &mut WriterCounter::new();
-        // these rules follow TX_MODE_LARGEST
```
Comment (ycho) [inline src/rdo.rs]: Ah, fixed here!

**Pull request merged**